Support v2 Linter API

vzamanillo opened the pull request:
Fixes #147

vzamanillo committed:
Support v2 Linter API

vzamanillo commented:
Ouch, whitequark/parser Warnings...

vzamanillo committed:
Fixed #117

We always handle two types of error, exec errors and program stream errors, in the first case we show a notification with the error related information, in the second case we check the exit code and text looking for warnings, if found we will show a warning notification with the warning related information, else an error notification. A new configuration option can suppress the warning notifications to avoid notification spam to the user. This will solve CI errors.

vzamanillo committed:
Fixed matching for multiple lines

vzamanillo reviewed (commented):
Fixed

Arcanemagus commented on lib/main.js:
```diff
@@ -60,7 +70,23 @@ export default {
+          return null;
+        }
+
+        if (output.exitCode !== 0) {
```
Is there an easy method to trigger this so it can be added to the specs?

vzamanillo replied:
I think using a ruby version with an old compilant syntax.

Arcanemagus replied:
Oh! This is the provider that had that, it's super easy to trigger when the CI version of Ruby is no longer supported by HAML, used to happen all the time.

That being said... the only way to test that would be a dedicated CI job so this is probably more complex than it's worth to implement it, I'm fine with leaving it without specs for now since the rest of the code isn't 100% tested in CI anyway.

vzamanillo replied:
Ok, we have time to think in how to do a spec.

jormon replied:
@Arcanemagus  - on my setup when any lints don't pass, the haml_lint returns with an exit code of 65, so this always triggers.  This is what's causing #169.  Are you familiar with what might be causing this?

Arcanemagus commented on lib/main.js:
```diff
@@ -72,7 +98,7 @@ export default {
 
         const messages = [];
-        let match = regex.exec(output);
+        let match = regex.exec(output.stdout);
```
Yep, that's how all of these work: It pulls the first match, then continues attempting to match on the remainder of the text until it runs out of text that matches.

Arcanemagus commented on package.json:
```diff
@@ -12,13 +12,18 @@
       "description": "Full path to a global Haml lint file, if no other is found",
       "type": "string"
+    },
+    "suppressWarnings": {
```
This could possibly be misconstrued to mean "Don't show me `warning` severity messages".

Maybe a better name would be `suppressFailures`? Is there any reason we don't just want to show all errors though? Generally that means something is wrong and the user isn't going to get any usable output otherwise.

vzamanillo replied:
What about changing the message to "Supress non-critical HAML-Lint messages" or a better way, adding a excerpt message to the results and show notification for errors only?.

vzamanillo replied:
Something like this

```
        const messages = [];

        if (output.exitCode !== 0) {
          if (output.stderr.toLowerCase().startsWith(warning)) {
            messages.push({
              severity: warning,
              excerpt: `HAML-Lint: ${output.stderr}`,
              location: {
                file: filePath,
                // first line of the file
                position: [[0, 0], [0, Infinity]],
              },
            });
          }
        } else {
          showErrorNotification('HAML-Lint: Unexpected error', output.stderr);
        }
```

Arcanemagus replied:
Adding it as a message looks great to me!

vzamanillo replied:
Should we add the unexpected errors this way too? I mean, without notification.

Arcanemagus replied:
The way I see it is:

- If the message prevents other results from being returned it should be an Atom error notification
- If other results can still be obtained from the linter then it probably should just be converted into a message on the entire first line

vzamanillo replied:
:+1:

vzamanillo committed:
Removed not needed html escape

vzamanillo committed:
First line message on haml-lint output warnings

suppressWarnings config. option and warning notification removed. If the message prevents other results from being returned it should be an Atom error notification, if other results can still be obtained from the linter then it probably should just be converted into a message on the entire first line

vzamanillo commented:
hehehe

`Expected 2 to be 1.`

If we modify the specs to match this we must ensure that the ruby versions of the CI have an old compilant syntax.

vzamanillo committed:
Bad condition

Error notification if not exit code 0 and is not a warning

vzamanillo committed:
Old compilant syntax ruby version support in specs

Conditional validation to pass tests with old compilant syntax ruby versions

Arcanemagus commented on lib/main.js:
```diff
@@ -60,7 +67,31 @@ export default {
+              },
+            });
+          } else {
+            showErrorNotification(output.stderr);
```
Shouldn't this `return null;` here as well to signify to the consumer that any current messages shouldn't be updated?

vzamanillo committed:
null return on unexpected errors

Arcanemagus reviewed (approved):
This looks good just looking at the code, is it ready to merge?

vzamanillo commented:
> This looks good just looking at the code, is it ready to merge?

Yep, it is ready!!
